datasets/flwr_datasets/partitioner/partitioner.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2023 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Partitioner class that works with HuggingFace Dataset."""
+from abc import ABC, abstractmethod
+from typing import Optional
+
+import datasets
+from datasets import Dataset
+
+
+class Partitioner(ABC):
+    """The base partitioner class that enables obtaining federated partitions.
+
+    The initialization is intended to take all necessary arguments such that the call to
+    the `load_partition` method can be use the same for all partitioners.
+    """
+
+    def __init__(self):
+        self._dataset: Optional[Dataset] = None
+
+    @property
+    def dataset(self):
+        """Dataset property."""
+        return self._dataset
+
+    @dataset.setter
+    def dataset(self, value: Dataset):
+        if self._dataset is None:
+            self._dataset = value
+        else:
+            raise Warning(
+                "The dataset should be assigned only once to the partitioner."
+                "This operation might also wipe out the saved references to the "
+                "created partitions (in case the partitioning scheme needs to create "
+                "the full partitioning also in order to return a single partition)."
+            )
```

Comment:
Sure, got it and fixed it now. Actually, it changes the logic here, but it changes to the correct flow - raise the warning but do the assignment anyway.

Reply:
@danieljanes just one question from me on that function. Later on, I think that the method is needed that will actually wipe out the reference if created. Sth like `_clear_reference` with the basic functionality that makes sure that the reference matches the dataset, and in case of the new dataset assignment, the new reference should be created. 
I hope it's fine if it's introduced later on bc there's no `reference` attribute currently and it should be needed from the PowerLaw/LDA Partitioning.
To clarify, by reference I mean the partition to index reference / alternatively already divided dataset so probably a dict - partition idx to partition data.
Does that make sense to you?